CONTRIBUTING.rst

```diff
@@ -147,20 +147,17 @@ If you plan on doing development and testing, you will also need to install the
 
 .. code-block:: bash
 
-    pip install -r dev-requirements.txt
-    pip install -r test-requirements.txt
+    pip install pip-tools
+    pip-compile --output-file compiled-requirements.txt dev-requirements.txt test-requirements.txt
```

Comment:
Is `pip-compile` strictly necessary for correctly installing dev and test requirements?

Reply:
We can likely do more user friendly. `pip-compile` is needed to avoid `pip` entering a near-infinite loop due to its package-conflict resolution logic. `pip-compile` does not need to be run by new users though. The latest guidance I am seeing from PyPI is to commit the compiled requirements file [[PyCon2021](https://www.youtube.com/watch?v=VWWgkF-0cDQ)][[pip-tools FAQ](https://github.com/jazzband/pip-tools#should-i-commit-requirementsin-and-requirementstxt-to-source-control)]. That ensures reproducible dev environments, reproducing builds, protects against package tampering, and removes the need for new users to use `pip-compile`. It does require existing committers to run `pip-compile` when updating the repo requirements, so it changes their workflow a bit. I could switch to committing the compiled requirements if this is fine with other committers.

Another improvement for new users could be to trim down the number of requirements to install upon first use. Version conflicts seem to primarily come from `extra-ml-requirements.txt`, which is likely not critical for a first commit doing a touch up in the code, unlike `lint-requirements.txt` which looks required to ensure success of a first PR.

Reply:
@maitre-matt Using `pip-compile` sounds good - thank you for the explanation. I also agree that taking `extra-ml-requirements` off the critical path for developer environment setup is a good idea.

Reply:
`extra-ml-requirements` is now a somewhat-outdated set of dependency recommendations for certain libraries as well. https://github.com/mlflow/mlflow/blob/master/mlflow/ml-package-versions.yml is the source of truth for our CI systems.